compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -1692,4 +1709,78 @@ impl<'context> Elaborator<'context> {
                 _ => true,
             })
     }
+
+    /// Check that a recursive function *can* return without endlessly calling itself.
+    fn check_for_unbounded_recursion(
+        &mut self,
+        func_id: FuncId,
+        func_name: String,
+        func_span: Span,
+        body_id: ExprId,
+    ) {
+        if !self.can_return_without_recursing(func_id, body_id) {
+            self.push_err(CompilationError::ResolverError(ResolverError::UnconditionalRecursion {
+                name: func_name,
+                span: func_span,
+            }));
+        }
+    }
+
+    /// Check if an expression will end up calling a specific function.
+    fn can_return_without_recursing(&self, func_id: FuncId, expr_id: ExprId) -> bool {
```

Comment:
I think this breaks in the scenario where you pass the current function as a function pointer into a call to a separate function. This other function could conditionally recuse but we treat it as unconditional.

There would likely be other cases with lambdas.

Reply:
Hm, do you have an example? We can add it to the tests. 

I'm not sure I got it right, but if it's a call to another function, it will immediately get a pass, ie. a false negative. For example mutual recursion is not rejected. 

I see I did `HirExpression::Lambda(e) => check(e.body)`, I suppose that might be wrong as it checks more than `HirExpression::Call` does, which doesn't follow into the body 🤔

Reply:
Added an exemption for lambda bodies. Rust doesn't seem to check them either.